---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---


2 Getting started

```{r}
#use these lines to install maSigPro
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("maSigPro")

#Use these 2 lines to view the guide
browseVignettes("maSigPro")
```

# START
```{r}
library(maSigPro)
help(package="maSigPro") #for package help
?p.vector #for function help
```


3 Multiple Series Time Course Experiment

# make.design.matrix(), p.vector(), T.fit(),
get.siggenes() and see.genes(). Additionally, the package provides the wrapping function
maSigPro() which executes the entire analysis in one go.

```{r}
data(data.abiotic)
data(edesign.abiotic)

# show table

data.abiotic
```


```{r}
colnames(data.abiotic)
"next"
"    "
"    "
rownames(edesign.abiotic)
"next"
"    "
"    "
colnames(edesign.abiotic)
"next"
"    "
"    "
rownames(data.abiotic)
```

3.1 Defining the regression model

```{r}
design <- make.design.matrix(edesign.abiotic, degree = 2)
#This example has three time points, so we can consider up to a quadratic regression model
#(degree = 2)
# ...matrix(data table , quadratic and etc)

# show design
design
```




```{r}
design$groups.vector # contains the assignment of regression variables to experimental groups.
```

3.2 Finding significant genes

```{r}

# do a regression fit
 fit <- p.vector(data.abiotic, design, Q = 0.05, MT.adjust = "BH", min.obs = 20)
 # the data you want to process
#     design matrix for the regression fit such as that generated by themake.design.matrix function
#                                  Q --> significance level

# MT.adjust ---> use BH meathod 
# min.obs --> genes with less than this number of true numerical values will be excluded from the analysis. Minimum value to estimate the model is (degree+1)xGroups+1. Default is 6.



#   performs a regression fit for each gene taking all variables present in the model given by a regression matrix and returns a list of FDR corrected significant genes.



fit$i # returns the number of significant genes
fit$alfa # gives p-value at the Q false discovery control level
fit$SELEC # is a matrix with the significant genes and their expression values





```

3.3 Finding significant differences

```{r}
tstep <- T.fit(fit, step.method = "backward", alfa = 0.05) # executes stepwise regression, d can be ”backward” or ”forward”
#T.fit(data, design = data$dis, step.method = "backward", 
#      min.obs = data$min.obs, alfa = data$Q....

```

3.4 Obtaining lists of significant genes



rsq: is a cutt-off value for the R-squared of the regression model.

vars: is used to indicate how to group variables to show results. There are 3 possible values:

groups: This will generate a list of significant genes for each experimental group.

The list corresponding to the reference group will contain genes whose expression
profile is significantly different from a 0 profile. The lists corresponding to the
remaining experimental groups will contain genes whose profiles are different from
the reference group.

all: One unique list of significant genes a any model variable will be produced.

each: There will be as many lists as variables in the regression model. This can be
used to analyze specific differences, for example genes that have linear or saturation
kinetics.

```{r}
sigs <- get.siggenes(tstep, rsq = 0.6, vars = "groups")
#( data, cut-off, vars name) --> we wraped all things together above, we can also use c(...) to group the vars we want
names(sigs)

# show names
names(sigs$sig.genes)
names(sigs$sig.genes$ColdvsControl)
```

4.1 Venn Diagrams

```{r}
G1 <-  suma2Venn(sigs$summary[, c(2:4)])# only show columns 2 to 4
G2 <-  suma2Venn(sigs$summary[, c(1:4)])# same rules


```


4.2 see.genes()
```{r}
sigs$sig.genes$SaltvsControl$g

#install.packages("nat")
# https://www.xquartz.org/
capabilities()
Sys.setenv("DISPLAY"=":0.0")


see.genes(sigs$sig.genes$ColdvsControl, show.fit = T, dis =design$dis,
cluster.method="hclust" ,cluster.data = 1, k = 9)
```


4.3 PlotGroups()

```{r}
STMDE66 <- data.abiotic[rownames(data.abiotic)=="STMDE66", ] # find the row called STMDE66

# check
STMDE66 # the row we want

PlotGroups (STMDE66, edesign = edesign.abiotic)

PlotGroups (STMDE66, edesign = edesign.abiotic, show.fit = T, dis = design$dis, 
            groups.vector = design$groups.vector)
#  row, data, show fitting line = True, dis = regression design matrix, groups.vec

```


4.4 PlotProfiles()
PlotProfiles(data, cond, main = NULL, cex.xaxis = 0.5, ylim = NULL, 
    repvect, sub = NULL, color.mode = "rainbow", item = NULL) 

```{r}
# EX

# PlotProfiles(edesign.abiotic, STMDE66, main = NULL, cex.xaxis = 0.5, ylim = NULL, 
#    repvect, sub = NULL, color.mode = "rainbow", item = NULL)

# We didn't assign repvect so we can't run this command successfully.
```



5.1 Single Series Time Course

```{r}
Time <- rep(c(1,5,10,24), each = 3) # rep replicates the values in x, each = Each element of x is repeated each times.
Replicates <- rep(c(1:4), each = 3)
Group <- rep(1,12)

ss.edesign <- cbind(Time,Replicates,Group) # columns bind
rownames(ss.edesign) <- paste("Array", c(1:12), sep = "") # paste several rows to ss.edesign names
```


```{r}
## Create data set

# writing function
# rnorm generates multivariate normal random variates in the space X.
#n    number of datasets to be simulated

#mean   The mean of the dataset to be simulated

#var    The variance covariance matrix

#lambda  The expected total count

#x   vectors in the sampling space

#withJacobian        should the jacobian of the log or logratio transformation be included in the density calculations? defaults to FALSE (see details)


ss.GENE <- function(n, r, var11 = 0.01, var12 = 0.02, var13 = 0.02,
var14 = 0.02, a1 = 0, a2 = 0, a3 = 0, a4 = 0) {
tc.dat <- NULL
for (i in 1:n) {
gene <- c(rnorm(r, a1, var11), rnorm(r, a1, var12),
rnorm(r, a3, var13), rnorm(r, a4, var14))
tc.dat <- rbind(tc.dat, gene)
}
tc.dat }
flat <-ss.GENE(n = 85, r = 3) # flat
induc <- ss.GENE(n = 5, r = 3, a1 = 0, a2 = 0.2, a3 = 0.6, a4 = 1) # induction
sat <- ss.GENE(n = 5, r = 3, a1 = 0, a2 = 1, a3 = 1.1, a4 = 1.2) # saturation
ord <- ss.GENE(n = 5, r = 3, a1 = -0.8, a2 = -1, a3 = -1.3, a4 =-0.9) # intercept
ss.DATA <- rbind(flat, induc,sat,ord)
rownames(ss.DATA) <- paste("feature", c(1:100), sep = "")
colnames(ss.DATA) <- paste("Array", c(1:12), sep = "")
# run maSigPro
ss.example <- maSigPro(ss.DATA, ss.edesign, vars="each")
```




5.2 Common Starting Time

```{r}
data(edesignCT)
```



6 Next Generation-Sequencing series

counts: a logical indicating whether your data are counts. By default is FALSE for microarray treatment.

theta: θ parameter for negative.binomial family. By default θ = 10.

family: the distribution function to be used in the GLM. It must be specified as a function: gaussian(), poisson(), negative.binomial(theta)... If NULL family will be negative.binomial(theta) when counts=TRUE or gaussian() when counts=FALSE.

```{r}
#NBdesign is the design matrix.
data(NBdata)
data(NBdesign)
d <- make.design.matrix(NBdesign)
#If we can use maSigPro with theta = 10:
library(MASS)
NBp <- p.vector(NBdata, d, counts=TRUE)
NBt <- T.fit(NBp)
get<- get.siggenes(NBt, vars="all")
get$summary


#see.genes(get$sig.genes, k = 4)

NBp <- p.vector(NBdata, d, counts=TRUE, theta=5)
NBp <- p.vector(NBdata, d, family=poisson() )
# family ---- the distribution function to be used in the glm model. It must be specified as a function: gaussian(), poisson(), negative.binomial(theta)... If NULL family will be negative.binomial(theta) when counts=TRUE or gaussian() when counts=FALSE



```


7 Iso-maSigPro: analysis of alternative isoform expression in time course transcriptomics experiments



```{r}
data(ISOdata)
data(ISOdesign)
dis <- make.design.matrix(ISOdesign)

```


7.1 IsoModel() and getDS()

IsoModel Performs a model comparison for each gene to detect genes with different trends in time course experiments and applies maSigPro to the Isoforms belonging to selected genes.

```{r}
MyIso <- IsoModel(data=ISOdata[,-1], gen=ISOdata[,1], design=dis, counts=TRUE)
# counts --- a logical indicating whether your data are counts

#design --- design matrix for the regression fit such as that generated by the make.design.matrix function
# data, 
```




```{r}
Myget <- getDS(MyIso)
# creates lists of significant isoforms from Differentially Spliced Genes (DSG)
# getDS(Model, vars="all", rsq=0.4)
#Model	---   a IsoModel object
# vars ---   argument of the get.siggenes function applied to isoforms
# rsq --- cut-off level at the R-squared value for the stepwise regression fit. Only isoforms with R-squared more than rsq are selected



Myget$DSG
Myget$DET
Myget$List0
```



7.2 Clustering strategy: seeDS() and tableDS()

```{r}

#see <- seeDS(Myget, cluster.all=FALSE, k=6)
# provides visualisation tools for Significant Isoforms in a time course experiment. 
# also needs see.genes()


# make it to table

#table <- tableDS(see)
#table$IsoTable

#getDSPatterns(table, 1, 4)
```


7.3 PodiumChange()

```{r}
PC<-PodiumChange(Myget, only.sig.iso=TRUE, comparison="specific",
group.name="Group2", time.points=c(18,24))
# provides lists of genes that have different Major isoforms (most expressed) when different intervals of the experimental conditions are considered.

#  only.sig.iso --- TRUE when changes are looked for only through statistically significant isoforms.
# comparison --- Type of search to do: any, groups or specific
PC$L
```

7.4 IsoPlot()

```{r}
IsoPlot(Myget,"Gene1005",only.sig.iso=FALSE,cex.main=2,cex.legend=1)
#This function makes a plot with the isoforms of a specific gene splitting the different experimental groups.

#Gene1005 --> the gene we want to show
# only.sig.iso ---	TRUE when the plot is made only with statistically significant isoforms.
# cex.main	--- graphical parameter magnification to be used for main
#cex.legend	--- graphical parameter magnification to be used for legend
```



















